service/src/main/java/bio/terra/axonserver/utils/CloudStorageUtils.java

```diff
@@ -70,15 +77,32 @@ public static byte[] getBucketObject(
             .getService()
             .reader(blobId)) {
 
+      long startByteIdx =
+          Optional.ofNullable(byteRange).isPresent() ? byteRange.getRangeStart(Long.MAX_VALUE) : 0;
+      long endByteIdx =
+          Optional.ofNullable(byteRange).isPresent()
+              ? byteRange.getRangeEnd(Long.MAX_VALUE)
+              : Long.MAX_VALUE - 1;
+      long bytesToRead = endByteIdx - startByteIdx;
+
       BoundedByteArrayOutputStream outputStream = new BoundedByteArrayOutputStream(MAX_OBJECT_SIZE);
-      ByteBuffer bytes = ByteBuffer.allocate(MAX_BUFFER_SIZE);
-      while (reader.read(bytes) > 0) {
-        bytes.flip();
-        outputStream.write(bytes.array(), 0, bytes.limit());
-        bytes.clear();
+      ByteBuffer buffer = ByteBuffer.allocate(MAX_BUFFER_SIZE);
```

Comment:
nit: given that we know `bytesToRead`, can we allocate a smaller buffer for smaller files (e.g. `.allocate(Math.min(MAX_BUFFER_SIZE, bytesToRead))`?

Reply:
Oh thats a good idea.. I ended up just adding a change that deals with a small file or range, but I think this is better.